Transaction termination aware locks

lutovich opened the pull request:
This PR makes single instance and HA locks transaction termination aware.
Feature is guarded by `unsupported.dbms.tx_termination_aware_locks` setting and off by default.
This is a reimplementation of https://github.com/neo4j/neo4j/pull/7168 for 2.3 branch.

lutovich committed:
Locks made transaction termination aware

KernelTransactionImplementation can be marked for termination from any thread
by calling #markForTermination(). This commit makes lock clients know that the
owning transaction was terminated so they can abort waiting.

lutovich committed:
Slave - master lock session termination

Commit makes it possible to terminate lock sessions that slaves have on master.
This is useful when terminating write transactions on slaves because they have
corresponding lock clients on master.

When SlaveLocksClient is stopped it will end the corresponding lock session on
master and stop local locks client.

Feature is guarded by `experimental.tx_termination_aware_locks` settings which
is off by default.

lutovich committed:
Improve integration testing of tx termination

All integration tests for transaction termination moved to a separate class in
'neo4j-integrationtests' module. This made test both community and forseti lock
managers.

lutovich committed:
Polish SlaveLocksClient#stop()

Commit makes SlaveLocksClient better handle stopping. It's methods throw
LockClientStoppedException when invoked on a stopped client.

Removed Locks.Client#releaseAll() because it was only used by SlaveLocksClient
and same functionality is available via Locks.Client#close(). Updated all
#releaseAll() tests respectively.

lutovich committed:
Added couple unit tests around tx termination

lutovich commented:
@digitalstain could you please take a look at this PR? it touches locking in HA.

digitalstain commented:
looks good to me

tinwelint commented:
@lutovich right off the bat I'd like to change `experimental.tx_termination_aware_locks` setting name to something that is compatible with 3.0-and-beyond naming scheme, to avoid different names between versions as much as possible. Something like `unsupported.dbms.tx_termination_aware_locks` comes to mind.

lutovich committed:
Rename tx_termination_aware_locks setting

Old name was `experimental.tx_termination_aware_locks` and new is
`unsupported.dbms.tx_termination_aware_locks` to be consistent with next
neo4j versions.

lutovich commented:
@tinwelint setting renamed

lutovich commented:
There are couple places where access to KTI state flags is racy. Fixing it.

lutovich committed:
Fix concurrency between KTI#markForTermination() and KTI#close()

It is possible that thread executing transaction termination runs concurrently
with thread that commits the transaction. Without proper synchronization
following issues could occur:
 * termination thread stops lock client that has been returned to the pool
   because tx was committed concurrently
 * termination thread termination thread marks wrong transaction for
   termination because it was committed and reused from the pool concurrently
 * two termination threads race and stop lock client twice

 This commit fixes the problem by introducing a lock between #release() and
 #markForTermination(). Later method also made smarter to check for logical
 transaction version, represented by reuseCount field.

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactionImplementation.java:
```diff
@@ -272,14 +286,42 @@ public boolean shouldBeTerminated()
+        }
+
+        int initialReuseCount = reuseCount;
+        terminationReleaseLock.lock();
```
What's the benefit we're getting from a lock object as opposed to using `synchronized` afaics there's no other synchronization in this class. It's explicit in what it does, sure, although it may be overkill a.t.m.

lutovich replied:
Having explicit lock or lock object just makes it a bit more clear what operations is it coordinating. In this case it's name `terminationReleaseLock` tries to explain smth. I can change this to be an `Object` and use `synchronized(terminationReleaseSync)` if needed.

tinwelint replied:
Ok I think that's fine and we're pooling the KTI instances anyway so really really low object overhead, I yield :)

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/Locks.java:
```diff
@@ -99,12 +99,10 @@ void visit( ResourceType resourceType, long resourceId, String description, long
         void releaseExclusive( ResourceType resourceType, long resourceId );
 
-        /** Release all locks. */
-        void releaseAll();
```
+1 this has had me confused from time to time

tinwelint commented on community/kernel/src/main/java/org/neo4j/kernel/impl/locking/LockClientStateHolder.java:
```diff
@@ -66,22 +66,23 @@ public void stopClient()
             if ( isStopped( currentState ) )
             {
-                return false;
+                throw new LockClientStoppedException( client );
```
So if we throw this exception then there's a chance user will see a `LockClientStoppedException` instead of `TransactionTerminatedException` as cause of tx failing, depending on which call happens to detect the fact that the tx has been terminated, will this be confusing to users?

lutovich replied:
`LockClientStoppedException` extends `TransactionTerminatedException` so if there is some automatic handling of `TransactionTerminatedException` than it'll work. Message of `LockClientStoppedException` will say `The transaction has been terminated. LockClient is stopped`

tinwelint replied:
Ok Great.

tinwelint commented on enterprise/ha/src/main/java/org/neo4j/kernel/ha/lock/SlaveLocksClient.java:
```diff
@@ -185,52 +198,58 @@ public void releaseExclusive( Locks.ResourceType resourceType, long resourceId )
+        }
+        else
+        {
+            throw new UnsupportedOperationException( "Lock client stop is unsupported on slave side." );
```
Wouldn't this be better as a no-op? I don't see how other lock clients throw exception if tx termination awareness is disabled.

lutovich replied:
`#stop()` was just originally throwing this exception. so if `!txTerminationAwareLocks` than same behaviour is preserved

tinwelint commented on enterprise/ha/src/test/java/org/neo4j/ha/ClusterTransactionIT.java:
```diff
@@ -55,8 +55,8 @@
         cluster = clusterRule.withProvider( clusterOfSize( 3 ) )
-                             .withSharedSetting( HaSettings.ha_server, ":6001-6005" )
-                             .withSharedSetting( HaSettings.tx_push_factor, "2" ).startCluster();
+                .withSharedSetting( HaSettings.ha_server, ":6001-6005" )
```
If we can avoid checkstyling in maint branches that'd be grand... never mind this change though.

lutovich committed:
Made SlaveLocksClient#stop() a noop when locks are termination aware

lutovich committed:
Make sure terminated transactions throw correct exceptions

Behaviour of KernelTransactionImplementation#close() is:
 * terminated transaction that was neither marked for success nor for failure
   simply rolls back without any exception
 * terminated transaction marked for success throws TransactionTerminatedException
 * terminated transaction marked for failure simply rolls back without any exception
 * terminated transaction marked for both success and failure throws
   TransactionTerminatedException

Added unit tests for all these cases.
